cmd/arrow/batches/arrow_batches.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"flag"
+	"fmt"
+	"github.com/apache/arrow/go/v12/arrow/array"
+	sf "github.com/snowflakedb/gosnowflake"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+)
+
+func getDSN() (string, *sf.Config, error) {
+	env := func(key string, failOnMissing bool) string {
+		if value := os.Getenv(key); key != "" {
+			return value
+		}
+		if failOnMissing {
+			log.Fatalf("%v environment variable is not set", key)
+		}
+		return ""
+	}
+
+	account := env("SNOWFLAKE_TEST_ACCOUNT", true)
+	user := env("SNOWFLAKE_TEST_USER", true)
+	password := env("SNOWFLAKE_TEST_PASSWORD", true)
+	host := env("SNOWFLAKE_TEST_HOST", false)
+	portStr := env("SNOWFLAKE_TEST_PORT", false)
+	protocol := env("SNOWFLAKE_TEST_PROTOCOL", false)
+
+	var err error
+	port := 443
+	if portStr != "" {
+		port, err = strconv.Atoi(portStr)
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	cfg := &sf.Config{
+		Account:  account,
+		User:     user,
+		Password: password,
+		Host:     host,
+		Port:     port,
+		Protocol: protocol,
+	}
+
+	dsn, err := sf.DSN(cfg)
+	return dsn, cfg, err
+}
+
+type sampleRecord struct {
+	batchID  int
+	workerID int
+	number   int32
+	string   string
+}
+
+func (s sampleRecord) String() string {
+	return fmt.Sprintf("batchID: %v, workerID: %v, number: %v, string: %v", s.batchID, s.workerID, s.number, s.string)
+}
+
+func main() {
+	if !flag.Parsed() {
+		flag.Parse()
+	}
+
+	dsn, cfg, err := getDSN()
+	if err != nil {
+		log.Fatalf("failed to create DSN from Config: %v, err: %v", cfg, err)
+	}
+
+	ctx := sf.WithArrowBatches(context.Background())
+	query := "SELECT SEQ4(), 'example ' || (SEQ4() * 2) FROM TABLE(GENERATOR(ROWCOUNT=>30000))"
+
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		log.Fatalf("failed to connect. %v, err: %v", dsn, err)
+	}
+	defer db.Close()
+
+	conn, _ := db.Conn(ctx)
+	var rows driver.Rows
+
+	err = conn.Raw(func(x interface{}) error {
+		rows, err = x.(driver.QueryerContext).QueryContext(ctx, query, nil)
+		return err
+	})
+	if err != nil {
+		log.Fatalf("unable to run the query. err: %v", err)
+	}
+	defer rows.Close()
+	defer conn.Close()
```

Comment:
shouldn't this be after line 87? Because it seems that it won't be registered when there is an error.

Reply:
You are perfectly right.